openeo/rest/job.py

```diff
@@ -424,6 +419,43 @@ def load_bytes(self) -> bytes:
     # TODO: more `load` methods e.g.: load GTiff asset directly as numpy array
 
 
+    def _download_to_file(self, url: str, target: Path, *, chunk_size: int=DEFAULT_DOWNLOAD_CHUNK_SIZE, range_size: int=DEFAULT_DOWNLOAD_RANGE_SIZE):
+        head = self.job.connection.head(url, stream=True)
+        if head.ok and head.headers.get("Accept-Ranges") == "bytes" and 'Content-Length' in head.headers:
+            file_size = int(head.headers['Content-Length'])
+            self._download_ranged(url=url, target=target, file_size=file_size, chunk_size=chunk_size, range_size=range_size)
+        else:
+            self._download_all_at_once(url=url, target=target, chunk_size=chunk_size)
+
+
+    def _download_ranged(self, url: str, target: Path, file_size: int, *, chunk_size: int=DEFAULT_DOWNLOAD_CHUNK_SIZE, range_size: int=DEFAULT_DOWNLOAD_RANGE_SIZE):
+        with target.open('wb') as f:
+            for from_byte_index in range(0, file_size, range_size):
+                to_byte_index = min(from_byte_index + range_size - 1, file_size - 1)
+                tries_left = MAX_RETRIES_PER_RANGE
+                while tries_left > 0:
+                    try:
+                        range_headers = {"Range": f"bytes={from_byte_index}-{to_byte_index}"}
+                        with self.job.connection.get(path=url, headers=range_headers, stream=True) as r:
+                            r.raise_for_status()
+                            shutil.copyfileobj(fsrc=r.raw, fdst=f, length=chunk_size)
+                        break
+                    except OpenEoApiPlainError as error:
+                        tries_left -= 1
+                        if tries_left > 0 and error.http_status_code in RETRIABLE_STATUSCODES:
+                            logger.warning(f"Failed to retrieve chunk {from_byte_index}-{to_byte_index} from {url} (status {error.http_status_code}) - retrying")
+                            continue
+                        else:
+                            raise error
+
+
+    def _download_all_at_once(self, url: str, target: Path, *, chunk_size: int=DEFAULT_DOWNLOAD_CHUNK_SIZE):
+        with self.job.connection.get(path=url, stream=True) as r:
+            r.raise_for_status()
+            with target.open("wb") as f:
+                shutil.copyfileobj(fsrc=r.raw, fdst=f, length=chunk_size)
```

Comment:
This should fall back to the original chunked implementation
```
        response = self._get_response(stream=True)
        with target.open("wb") as f:
            for block in response.iter_content(chunk_size=chunk_size):
                f.write(block)
```

because this change loads in all data in memory, which is a regression

Reply:
shutil.copyfileobj(fsrc=r.raw, fdst=f, length=chunk_size) performs a chunked copy (https://docs.python.org/3/library/shutil.html#:~:text=shutil.copyfileobj,will%20be%20copied.)